### PR TITLE
Use all test data

### LINKED ISSUE #27: [Task]: use all test data for unit tests
test data are stored in pcse repo under the [folder tests](https://github.com/ajwdewit/pcse/tree/master/tests/test_data). We have to use all of them in unit tests because only using one test data might not be enough to test all conditions in the crop module, e.g. in #12. 

It would be best if pytest can download the data directly from the repo, instead of copying those files to this repo.

### PR DESCRIPTION
**Pull request by SarahAlidoost**
closes #27 

This PR:
- reads test data directly from its github repo, no need to download them
- uses all data in tests for checking the correctness of the model
- fixes a few bugs in leaf dynamics

### CODE REVIEW

File: src/diffwofost/physical_models/crop/leaf_dynamics.py [hunk 1]
```diff
@@ -232,68 +232,97 @@ def calc_rates(self, day: datetime.date, drv: WeatherDataContainer) -> None:
...
+        # the if statement `p.SPAN.requires_grad` to avoid unnecessary
+        # approximation when SPAN is not a learnable parameter.
+        # TODO: sharpness can be exposed as a parameter
+        if p.SPAN.requires_grad:
+            sharpness = torch.tensor(1000.0, dtype=DTYPE)
+            weight = torch.sigmoid((s.LVAGE - tSPAN) * sharpness)
+            r.DALV = torch.sum(weight * s.LV, dim=-1)
+        else:
+            span_mask = (s.LVAGE > tSPAN).to(dtype=DTYPE)
+            r.DALV = torch.sum(span_mask * s.LV, dim=-1)
```
Comment (fnattino) [inline src/diffwofost/physical_models/crop/leaf_dynamics.py]: Ah, it's probably me who removed this in the previous iteration. I think I only now understand why you had the if-block here: you want the code to return identical results to the original PCSE when SPAN is not a learnable parameter, while you want to use the approach with the sigmoid when you want to track the gradient - is this correct? 

I wonder whether this approach would not create "surprises" at a later stage. Wouldn't be better to have diffwofost to consistently return leaf dynamics results calculated using the sigmoid approach, even if they slightly differ from the original WOFOST implementation? 

I am thinking e.g. to the situation when you train a model (thus using the sigmoid-based approach) and then you run inference with the model (thus using the approach that is more similar to the original WOFOST). The implementation used in the two tasks would be slightly different and might produce slightly different results?

File: src/diffwofost/physical_models/crop/leaf_dynamics.py [hunk 2]
```diff
@@ -232,68 +232,97 @@ def calc_rates(self, day: datetime.date, drv: WeatherDataContainer) -> None:
...
+            torch.where(is_lai_exp, r.GRLV * r.SLAT, r.GLASOL),
+            torch.tensor(0.0, dtype=DTYPE),
+        )
+
         # sink-limited increase in leaf area
         GLA = torch.minimum(r.GLAIEX, r.GLASOL)
+
         # adjustment of specific leaf area of youngest leaf class
-        r.SLAT = torch.where(is_lai_exp & (r.GRLV > 0.0), GLA / r.GRLV, r.SLAT)
+        r.SLAT = torch.where(
```
Comment (fnattino) [inline src/diffwofost/physical_models/crop/leaf_dynamics.py]: Same as above for `r.GLAIEX`

File: src/diffwofost/physical_models/crop/leaf_dynamics.py (hunk 1 above)
Comment (SarahAlidoost) [reply src/diffwofost/physical_models/crop/leaf_dynamics.py]: > Ah, it's probably me who removed this in the previous iteration. I think I only now understand why you had the if-block here: you want the code to return identical results to the original PCSE when SPAN is not a learnable parameter, while you want to use the approach with the sigmoid when you want to track the gradient - is this correct?

that's correct. 

> 
> I wonder whether this approach would not create "surprises" at a later stage. Wouldn't be better to have diffwofost to consistently return leaf dynamics results calculated using the sigmoid approach, even if they slightly differ from the original WOFOST implementation?

Thanks for the good questions. The answer is that we don’t get any big “surprises,” since the sigmoid function provides smooth, soft transitions. But the amount of approximation depends on the `sharpness` value, which can be a learnable parameter. For now, I’ve hardcoded the values based on some experiments with the test data. To show this, I added the test `test_leaf_dynamics_with_sigmoid_approx.` In general, we should be able to disable these approximations when needed. So, I’d suggest keeping the if condition `if p.SPAN.requires_grad`. 

These kinds of approximations are normal when making physical models differentiable, hard thresholds become soft ones, which will slightly change the results. That’s inevitable. At this point, we don’t know yet how many of these approximations exist in pcse models, maybe this is the only one, and it just has a single parameter. We also don’t yet know how many use cases actually depend on the gradient of `SPAN`. A better solution would be to define how much tolerance we can accept when comparing against pcse results. I mentioned this before in issue #24, and we’ll discuss this again with the research team. 

> 
> I am thinking e.g. to the situation when you train a model (thus using the sigmoid-based approach) and then you run inference with the model (thus using the approach that is more similar to the original WOFOST). The implementation used in the two tasks would be slightly different and might produce slightly different results?

No, during inference, we run the model inside a `with torch.no_grad()` block. This doesn’t change `p.SPAN.requires_grad`, so the sigmoid code is used in both training and inference, see [torch examples, especially the last one](https://docs.pytorch.org/docs/stable/generated/torch.no_grad.html). The code `with torch.no_grad()` disables autograd tracking globally, meaning the computation graph isn’t built and `p.SPAN.requires_grad` remains `True` unless the user sets `p.SPAN.requires_grad = False`.

**Review by fnattino — APPROVED** [review]
All looks good, thanks for the clarifications!
